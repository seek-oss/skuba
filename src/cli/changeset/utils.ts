// Adapted from https://github.com/changesets/action/blob/21240c3cd1d2efa2672d64e0235a03cf139b83e6/src/gitUtils.ts

import type { Package } from '@manypkg/get-packages';
import { getPackages } from '@manypkg/get-packages';
import execa from 'execa';
import mdastToString from 'mdast-util-to-string';
import remarkParse from 'remark-parse';
import remarkStringify from 'remark-stringify';
import { unified } from 'unified';

export const BumpLevels = {
  dep: 0,
  patch: 1,
  minor: 2,
  major: 3,
} as const;

export async function getVersionsByDirectory(cwd: string) {
  const { packages } = await getPackages(cwd);
  return new Map(packages.map((x) => [x.dir, x.packageJson.version]));
}

export async function getChangedPackages(
  cwd: string,
  previousVersions: Map<string, string>,
) {
  const { packages } = await getPackages(cwd);
  const changedPackages = new Set<Package>();

  for (const pkg of packages) {
    const previousVersion = previousVersions.get(pkg.dir);
    if (previousVersion !== pkg.packageJson.version) {
      changedPackages.add(pkg);
    }
  }

  return [...changedPackages];
}

export function getChangelogEntry(changelog: string, version: string) {
  const ast = unified().use(remarkParse).parse(changelog);

  let highestLevel: number = BumpLevels.dep;

  const nodes = ast.children;
  let headingStartInfo:
    | {
        index: number;
        depth: number;
      }
    | undefined;
  let endIndex: number | undefined;

  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i];
    if (node.type === 'heading') {
      const stringified: string = mdastToString.toString(node);
      const match = /(major|minor|patch)/.exec(stringified.toLowerCase());
      if (match !== null) {
        const level = BumpLevels[match[0] as 'major' | 'minor' | 'patch'];
        highestLevel = Math.max(level, highestLevel);
      }
      if (headingStartInfo === undefined && stringified === version) {
        headingStartInfo = {
          index: i,
          depth: node.depth,
        };
        continue;
      }
      if (
        endIndex === undefined &&
        headingStartInfo !== undefined &&
        headingStartInfo.depth === node.depth
      ) {
        endIndex = i;
        break;
      }
    }
  }
  if (headingStartInfo) {
    ast.children = ast.children.slice(headingStartInfo.index + 1, endIndex);
  }
  return {
    content: unified().use(remarkStringify).stringify(ast),
    highestLevel,
  };
}

export const execWithOutput = async (
  command: string,
  args?: string[],
  options: { ignoreReturnCode?: boolean; cwd?: string } = {},
) => {
  try {
    const { exitCode, stdout, stderr } = await execa(
      command,
      args,
      ...(options.cwd ? [{ cwd: options.cwd }] : []),
    );

    return {
      code: exitCode,
      stdout,
      stderr,
    };
  } catch ({ exitCode, stdout, stderr }) {
    if (!options.ignoreReturnCode) {
      throw new Error(stderr as string);
    }

    return {
      code: exitCode,
      stdout,
      stderr,
    };
  }
};

export function sortTheThings(
  a: { private: boolean; highestLevel: number },
  b: { private: boolean; highestLevel: number },
) {
  if (a.private === b.private) {
    return b.highestLevel - a.highestLevel;
  }
  if (a.private) {
    return 1;
  }
  return -1;
}
