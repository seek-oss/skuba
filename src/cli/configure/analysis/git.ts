import fs from 'fs-extra';
import git from 'isomorphic-git';

import { crawlDirectory } from '../../../utils/dir';
import { log } from '../../../utils/logging';

export const auditWorkingTree = async (dir: string) => {
  const filepaths = await crawlDirectory(dir);

  const statuses = await Promise.all(
    filepaths.map((filepath) => git.status({ dir, fs, filepath })),
  );

  if (
    statuses.some(
      (status) =>
        status !== 'absent' && status !== 'ignored' && status !== 'unmodified',
    )
  ) {
    log.newline();
    log.warn('You have dirty/untracked files that may be overwritten.');
  }
};
