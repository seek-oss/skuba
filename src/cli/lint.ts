import { execConcurrently } from '../utils/exec';

export const lint = () =>
  execConcurrently([
    {
      command: 'eslint --ext=js,ts,tsx .',
      name: 'ESLint',
    },
    {
      command: 'prettier --check .',
      name: 'Prettier',
    },
    {
      command: 'tsc --noEmit',
      name: 'tsc',
    },
  ]);
