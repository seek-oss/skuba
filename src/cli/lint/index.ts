import { hasDebugFlag, hasSerialFlag } from '../../utils/args';

import { externalLint } from './external';
import { internalLint } from './internal';
import type { Input } from './types';

export const lint = async (
  args = process.argv,
  tscOutputStream: NodeJS.WritableStream | undefined = undefined,
) => {
  const opts: Input = {
    debug: hasDebugFlag(args),
    serial: hasSerialFlag(args),
    tscOutputStream,
  };

  await externalLint(opts);

  await internalLint();
};
