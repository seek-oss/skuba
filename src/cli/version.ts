import { getSkubaVersion } from '../utils/version';

export const version = async () => {
  const skubaVersion = await getSkubaVersion();

  console.log(skubaVersion);
};
