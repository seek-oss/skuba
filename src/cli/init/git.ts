import path from 'path';

import fs from 'fs-extra';
import git from 'isomorphic-git';
import simpleGit from 'simple-git';

import * as Git from '../../api/git';
import { log } from '../../utils/logging';

interface GitHubProject {
  orgName: string;
  repoName: string;
  defaultBranch: string;
}

export const initialiseRepo = async (
  dir: string,
  { orgName, repoName, defaultBranch }: GitHubProject,
) => {
  await git.init({
    defaultBranch,
    dir,
    fs,
  });

  await Git.commit({
    dir,
    message: 'Initial commit',
  });

  await git.addRemote({
    dir,
    fs,
    remote: 'origin',
    url: `git@github.com:${orgName}/${repoName}.git`,
  });
};

export const downloadGitHubTemplate = async (
  gitHubPath: string,
  destinationDir: string,
) => {
  log.newline();
  log.plain('Downloading', log.bold(gitHubPath), 'from GitHub...');

  await simpleGit().clone(`git@github.com:${gitHubPath}.git`, destinationDir, [
    '--depth=1',
    '--quiet',
  ]);

  await fs.promises.rm(path.join(destinationDir, '.git'), {
    force: true,
    recursive: true,
  });

  log.newline();
  log.warn(
    'You may need to run',
    log.bold('yarn skuba configure'),
    'once this is done.',
  );
};
