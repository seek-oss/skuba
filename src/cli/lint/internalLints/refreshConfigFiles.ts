import path from 'path';
import { inspect } from 'util';

import { writeFile } from 'fs-extra';
import stripAnsi from 'strip-ansi';

import { Git } from '../../..';
import type { Logger } from '../../../utils/logging';
import { NPMRC_LINES, hasNpmrcSecret } from '../../../utils/npmrc';
import {
  type PackageManagerConfig,
  detectPackageManager,
} from '../../../utils/packageManager';
import { readBaseTemplateFile } from '../../../utils/template';
import { getDestinationManifest } from '../../configure/analysis/package';
import { createDestinationFileReader } from '../../configure/analysis/project';
import { mergeWithConfigFile } from '../../configure/processing/configFile';
import type { InternalLintResult } from '../internal';

const ensureNoAuthToken = (fileContents: string) =>
  fileContents
    .split('\n')
    .filter((line) => !hasNpmrcSecret(line))
    .join('\n');

type RefreshableConfigFile = {
  name: string;
  type: 'ignore' | 'npmrc';
  additionalMapping?: (
    s: string,
    packageManager: PackageManagerConfig,
  ) => string;
  if?: (packageManager: PackageManagerConfig) => boolean;
};

const removeRedundantNpmrc = (contents: string) => {
  const npmrcLines = contents
    .split('\n')
    .filter((line) => NPMRC_LINES.includes(line.trim()));

  // If we're only left with !.npmrc line we can remove it
  // TODO: Consider if we should generalise this
  if (npmrcLines.length > 0 && npmrcLines.every((line) => line.includes('!'))) {
    return contents
      .split('\n')
      .filter((line) => !NPMRC_LINES.includes(line.trim()))
      .join('\n');
  }
  return contents;
};

export const REFRESHABLE_CONFIG_FILES: RefreshableConfigFile[] = [
  { name: '.eslintignore', type: 'ignore' },
  {
    name: '.gitignore',
    type: 'ignore',
    additionalMapping: removeRedundantNpmrc,
  },
  { name: '.prettierignore', type: 'ignore' },
  {
    name: '.npmrc',
    type: 'npmrc',
    additionalMapping: ensureNoAuthToken,
    if: (packageManager: PackageManagerConfig) =>
      packageManager.command === 'pnpm',
  },
  {
    name: '.dockerignore',
    type: 'ignore',
    additionalMapping: removeRedundantNpmrc,
  },
];

export const refreshConfigFiles = async (
  mode: 'format' | 'lint',
  logger: Logger,
) => {
  const [manifest, gitRoot] = await Promise.all([
    getDestinationManifest(),
    Git.findRoot({ dir: process.cwd() }),
  ]);

  const destinationRoot = path.dirname(manifest.path);

  const readDestinationFile = createDestinationFileReader(destinationRoot);

  const refreshConfigFile = async (
    {
      name: filename,
      type: fileType,
      additionalMapping = (s) => s,
      if: condition = () => true,
    }: RefreshableConfigFile,
    packageManager: PackageManagerConfig,
  ) => {
    if (!condition(packageManager)) {
      return { needsChange: false };
    }

    const [inputFile, templateFile, isGitIgnored] = await Promise.all([
      readDestinationFile(filename),
      readBaseTemplateFile(`_${filename}`),
      gitRoot
        ? Git.isFileGitIgnored({
            gitRoot,
            absolutePath: path.join(destinationRoot, filename),
          })
        : false,
    ]);

    // If the file is gitignored and doesn't exist, don't make it
    if (inputFile === undefined && isGitIgnored) {
      return { needsChange: false };
    }

    const data = additionalMapping(
      inputFile
        ? mergeWithConfigFile(templateFile, fileType)(inputFile)
        : templateFile,
      packageManager,
    );

    const filepath = path.join(destinationRoot, filename);

    if (mode === 'format') {
      if (data === inputFile) {
        return { needsChange: false };
      }

      await writeFile(filepath, data);
      return {
        needsChange: false,
        msg: `Refreshed ${logger.bold(filename)}.`,
        filename,
      };
    }

    if (data !== inputFile) {
      return {
        needsChange: true,
        msg: `The ${logger.bold(
          filename,
        )} file is out of date. Run \`${logger.bold(
          packageManager.exec,
          'skuba',
          'format',
        )}\` to update it.`,
        filename,
      };
    }

    return { needsChange: false };
  };

  const packageManager = await detectPackageManager(destinationRoot);

  const results = await Promise.all(
    REFRESHABLE_CONFIG_FILES.map((conf) =>
      refreshConfigFile(conf, packageManager),
    ),
  );

  // Log after for reproducible test output ordering
  results.forEach((result) => {
    if (result.msg) {
      logger.warn(result.msg, logger.dim('refresh-config-files'));
    }
  });

  const anyNeedChanging = results.some(({ needsChange }) => needsChange);

  return {
    ok: !anyNeedChanging,
    fixable: anyNeedChanging,
    annotations: results.flatMap(({ needsChange, filename, msg }) =>
      needsChange && msg
        ? [
            {
              path: filename,
              message: stripAnsi(msg),
            },
          ]
        : [],
    ),
  };
};

export const tryRefreshConfigFiles = async (
  mode: 'format' | 'lint',
  logger: Logger,
): Promise<InternalLintResult> => {
  try {
    return await refreshConfigFiles(mode, logger);
  } catch (err) {
    logger.warn('Failed to refresh config files.');
    logger.subtle(inspect(err));

    return {
      ok: false,
      fixable: false,
      annotations: [],
    };
  }
};
