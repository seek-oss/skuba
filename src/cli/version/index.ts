import { log } from '../../utils/logging.js';
import { getSkubaVersion } from '../../utils/version.js';

export const version = async () => {
  const skubaVersion = await getSkubaVersion();

  log.plain(skubaVersion);
};
