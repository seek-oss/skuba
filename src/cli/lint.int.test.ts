import crypto from 'crypto';
import path from 'path';
import stream from 'stream';
import { inspect } from 'util';

import fs, { copy } from 'fs-extra';
import git from 'isomorphic-git';

import { Buildkite } from '..';
import type { Logger } from '../utils/logging';
import { getSkubaVersion } from '../utils/version';

import { lint } from './lint';
import { refreshConfigFiles } from './lint/internalLints/refreshConfigFiles';

jest.setTimeout(30_000);

jest.mock('../utils/version');

const buildkiteAnnotate = jest.spyOn(Buildkite, 'annotate').mockResolvedValue();

const stdoutMock = jest.fn();

jest
  .spyOn(console, 'log')
  .mockImplementation((...args) => stdoutMock(`${args.join(' ')}\n`));

jest
  .spyOn(git, 'listRemotes')
  .mockResolvedValue([
    { remote: 'origin', url: 'git@github.com:seek-oss/skuba.git' },
  ]);

const tscOutputStream = new stream.PassThrough().on('data', stdoutMock);

const BASE_PATH = path.join(__dirname, '..', '..', 'integration', 'base');

const TEMP_PATH = path.join(__dirname, '..', '..', 'integration', 'lint');

const stdout = (randomMatcher: RegExp) => {
  const result = stdoutMock.mock.calls
    .flat(1)
    .join('')
    .replace(/ in [\d\.]+s\./g, ' in <random>s.')
    .replace(
      /tsc      │ Lines of ([^:]+):[ ]+\d+/g,
      'tsc      │ Lines of $1: <random>',
    )
    .replace(
      /tsc      │ Nodes of ([^:]+):[ ]+\d+/g,
      'tsc      │ Nodes of $1: <random>',
    )
    .replace(
      /tsc      │ (Files|Identifiers|Symbols|Types|Instantiations|Memory used):[ ]+\d+/g,
      'tsc      │ $1: <random>',
    )
    .replace(
      /tsc      │ (.+) cache size:[ ]+\d+/g,
      'tsc      │ $1 cache size: <random>',
    )
    .replace(
      /tsc      │ (.+) time:[ ]+[\d\.]+s/g,
      'tsc      │ $1 time: <random>s',
    )
    .replace(randomMatcher, '<random>');

  return `\n${result}`;
};

const prepareTempDirectory = async (baseDir: string, tempDir: string) => {
  await copy(baseDir, tempDir);
  process.chdir(tempDir);
  await git.init({ fs, dir: tempDir });
  const result = await refreshConfigFiles('format', {
    bold: jest.fn(),
    dim: jest.fn(),
    warn: jest.fn(),
  } as unknown as Logger);
  expect(result.ok).toBe(true);
};

const originalCwd = process.cwd();

beforeAll(() => {
  delete process.env.BUILDKITE;
  delete process.env.CI;
  delete process.env.GITHUB_ACTIONS;
});

beforeEach(() => {
  jest.clearAllMocks();

  process.exitCode = undefined;
});

afterAll(() => {
  process.exitCode = undefined;

  // Restore the original working directory to avoid confusion in other tests.
  process.chdir(originalCwd);

  // Clean up temporary directories to avoid subsequent formatting and linting
  // warnings and to save on disk space. This can be commented out to inspect
  // the output.
  return fs.promises.rm(TEMP_PATH, {
    force: true,
    recursive: true,
  });
});

interface Args {
  args: string[];
  base: string;
  skubaVersion: string;
  exitCode: number | undefined;
}

test.each`
  description        | args           | base           | skubaVersion | exitCode
  ${'fixable'}       | ${[]}          | ${'fixable'}   | ${'0.0.0'}   | ${1}
  ${'ok'}            | ${[]}          | ${'ok'}        | ${'0.0.0'}   | ${undefined}
  ${'ok --debug'}    | ${['--debug']} | ${'ok'}        | ${'0.0.0'}   | ${undefined}
  ${'unfixable'}     | ${[]}          | ${'unfixable'} | ${'0.0.0'}   | ${1}
  ${'needs patches'} | ${[]}          | ${'patch'}     | ${'1.0.0'}   | ${1}
`('$description', async ({ args, base, skubaVersion, exitCode }: Args) => {
  jest.mocked(getSkubaVersion).mockResolvedValue(skubaVersion);

  const baseDir = path.join(BASE_PATH, base);

  const tempDir = path.join(
    TEMP_PATH,
    `${base}-${crypto.randomBytes(32).toString('hex')}`,
  );

  await prepareTempDirectory(baseDir, tempDir);

  await expect(
    // Disable worker threads as they can't run in a TypeScript context.
    lint(args, tscOutputStream, false),
  ).resolves.toBeUndefined();

  const tempDirRegex = new RegExp(tempDir, 'g');

  expect(stdout(tempDirRegex)).toMatchSnapshot();

  expect(
    buildkiteAnnotate.mock.calls.map(
      ([markdown, opts]) =>
        `\nOptions: ${inspect(opts)}\n\n${markdown.replace(
          tempDirRegex,
          '<random>',
        )}\n`,
    ),
  ).toMatchSnapshot();

  expect(process.exitCode).toBe(exitCode);
});
