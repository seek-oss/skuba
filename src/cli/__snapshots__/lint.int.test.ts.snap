// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fixable 1`] = `
"
ESLint   │ Processed 4 files in <random>s.
ESLint   │ 
<random>/a/a/a.mjs
  3:1  error  \`fs\` import should occur before import of \`path\`  import/order

<random>/a/a/a.mts
  3:1  error  \`fs\` import should occur before import of \`path\`  import/order

<random>/a/a/a.ts
  3:1  error  \`fs\` import should occur before import of \`path\`  import/order

✖ 3 problems (3 errors, 0 warnings)
  3 errors and 0 warnings potentially fixable with the \`--fix\` option.

Prettier │ Processed 8 files in <random>s.
Prettier │ Flagged 4 files:
Prettier │ b.md
Prettier │ c.json
Prettier │ d.js
Prettier │ package.json
tsc      │ TSFILE: <random>/lib/tsconfig.tsbuildinfo
tsc      │ tsc --noEmit exited with code 0
skuba    │ Processed skuba lints in <random>s.
ESLint, Prettier found issues that require triage.

Try running pnpm exec skuba format to fix them.
"
`;

exports[`fixable 2`] = `
[
  "
Options: { context: 'skuba-lint', scopeContextToStep: true, style: 'error' }

\`skuba lint\` found issues that require triage:

**ESLint**

\`\`\`term
<random>/a/a/a.mjs
  3:1  error  \`fs\` import should occur before import of \`path\`  import/order

<random>/a/a/a.mts
  3:1  error  \`fs\` import should occur before import of \`path\`  import/order

<random>/a/a/a.ts
  3:1  error  \`fs\` import should occur before import of \`path\`  import/order

✖ 3 problems (3 errors, 0 warnings)
  3 errors and 0 warnings potentially fixable with the \`--fix\` option.
\`\`\`

**Prettier**

\`\`\`term
b.md
c.json
d.js
package.json
\`\`\`
",
]
`;

exports[`needs patches 1`] = `
"
ESLint   │ Processed 2 files in <random>s.
Prettier │ Processed 6 files in <random>s.
tsc      │ TSFILE: <random>/lib/tsconfig.tsbuildinfo
tsc      │ tsc --noEmit exited with code 0
skuba    │ skuba has patches to apply. Run pnpm exec skuba format to run them. skuba-patches
skuba    │ Processed skuba lints in <random>s.
skuba found issues that require triage.

Try running pnpm exec skuba format to fix them.
"
`;

exports[`needs patches 2`] = `
[
  "
Options: { context: 'skuba-lint', scopeContextToStep: true, style: 'error' }

\`skuba lint\` found issues that require triage:

**skuba**

\`\`\`term
<random>/package.json skuba has patches to apply. Run pnpm exec skuba format to run them.
\`\`\`
",
]
`;

exports[`ok --debug 1`] = `
"
ESLint   │ Initialising ESLint...
ESLint   │ Processing files...
ESLint   │ Processed 2 files in <random>s.
ESLint   │ ○ a/a/a.ts
ESLint   │ ○ d.js
Prettier │ Initialising Prettier...
Prettier │ Detected project root: <random>
Prettier │ Discovering files...
Prettier │ Discovered 11 files.
Prettier │ Linting files...
Prettier │ .dockerignore
Prettier │   parser: -
Prettier │ .eslintignore
Prettier │   parser: -
Prettier │ .gitignore
Prettier │   parser: -
Prettier │ .npmrc
Prettier │   parser: -
Prettier │ .prettierignore
Prettier │   parser: -
Prettier │ b.md
Prettier │   parser: markdown
Prettier │ c.json
Prettier │   parser: json
Prettier │ d.js
Prettier │   parser: babel
Prettier │ package.json
Prettier │   parser: json-stringify
Prettier │ tsconfig.json
Prettier │   parser: json
Prettier │ a/a/a.ts
Prettier │   parser: typescript
Prettier │ Processed 6 files in <random>s.
tsc      │ TSFILE: <random>/lib/tsconfig.tsbuildinfo
tsc      │ Files: <random>
tsc      │ Lines of Library: <random>
tsc      │ Lines of Definitions: <random>
tsc      │ Lines of TypeScript: <random>
tsc      │ Lines of JavaScript: <random>
tsc      │ Lines of JSON: <random>
tsc      │ Lines of Other: <random>
tsc      │ Identifiers: <random>
tsc      │ Symbols: <random>
tsc      │ Types: <random>
tsc      │ Instantiations: <random>
tsc      │ Memory used: <random>K
tsc      │ Assignability cache size: <random>
tsc      │ Identity cache size: <random>
tsc      │ Subtype cache size: <random>
tsc      │ Strict subtype cache size: <random>
tsc      │ I/O Read time: <random>s
tsc      │ Parse time: <random>s
tsc      │ ResolveModule time: <random>s
tsc      │ ResolveTypeReference time: <random>s
tsc      │ ResolveLibrary time: <random>s
tsc      │ Program time: <random>s
tsc      │ Bind time: <random>s
tsc      │ Check time: <random>s
tsc      │ I/O Write time: <random>s
tsc      │ printTime time: <random>s
tsc      │ Emit time: <random>s
tsc      │ Total time: <random>s
tsc      │ tsc --extendedDiagnostics --noEmit exited with code 0
skuba    │ Processed skuba lints in <random>s.
"
`;

exports[`ok --debug 2`] = `[]`;

exports[`ok 1`] = `
"
ESLint   │ Processed 2 files in <random>s.
Prettier │ Processed 6 files in <random>s.
tsc      │ TSFILE: <random>/lib/tsconfig.tsbuildinfo
tsc      │ tsc --noEmit exited with code 0
skuba    │ Processed skuba lints in <random>s.
"
`;

exports[`ok 2`] = `[]`;

exports[`unfixable 1`] = `
"
ESLint   │ Processed 2 files in <random>s.
ESLint   │ 
<random>/a/a/a.ts
  3:1  error  \`fs\` import should occur before import of \`path\`  import/order
  8:3  error  Unexpected console statement                      no-console

<random>/d.js
  1:0  error  Parsing error: Unexpected token (1:0)

✖ 3 problems (3 errors, 0 warnings)
  1 error and 0 warnings potentially fixable with the \`--fix\` option.

Prettier │ Processed 6 files in <random>s.
Prettier │ Flagged 2 files:
Prettier │ d.js SyntaxError: Unexpected token (1:1)
> 1 | }
    | ^
  2 |
Prettier │ a/a/a.ts
tsc      │ d.js(1,1): error TS1128: Declaration or statement expected.
tsc      │ TSFILE: <random>/lib/tsconfig.tsbuildinfo
tsc      │ tsc --noEmit exited with code 2
skuba    │ Processed skuba lints in <random>s.
ESLint, Prettier, tsc found issues that require triage.

Try running pnpm exec skuba format to fix them.
"
`;

exports[`unfixable 2`] = `
[
  "
Options: { context: 'skuba-lint', scopeContextToStep: true, style: 'error' }

\`skuba lint\` found issues that require triage:

**ESLint**

\`\`\`term
<random>/a/a/a.ts
  3:1  error  \`fs\` import should occur before import of \`path\`  import/order
  8:3  error  Unexpected console statement                      no-console

<random>/d.js
  1:0  error  Parsing error: Unexpected token (1:0)

✖ 3 problems (3 errors, 0 warnings)
  1 error and 0 warnings potentially fixable with the \`--fix\` option.
\`\`\`

**Prettier**

\`\`\`term
d.js SyntaxError: Unexpected token (1:1)
> 1 | }
    | ^
  2 |
a/a/a.ts
\`\`\`

**tsc**

\`\`\`term
d.js(1,1): error TS1128: Declaration or statement expected.
tsc --noEmit exited with code 2
\`\`\`
",
]
`;
